### ISSUE TITLE
redeploying a plugin prevents event handler initialization

### ISSUE THREAD

**Issue by damien-git** [blueriver/MuraCMS#2630]
I had a "bean not found" error in a plugin. The beans were registered in the applicationLoad event, and I realized it was not called anymore. Looking further into this issue, I noticed that the rsScripts1 and rsScripts2 queries in pluginManager.loadPlugins were only keeping event handlers for plugins with deployed=1, and redeploying a plugin causes deployed to be set to 2. It looks like a bug (using latest Mura7)...

**Comment by David-Polehonski** [comment]
Would it be possible to get a an excerpt from the exception log, and if possible a reproducable test case?

**Comment by damien-git** [comment]
I don't remember how I ran into this issue, so I am not sure it is reproducable.

But more recently I noticed that when I update a plugin by uploading a ZIP and clicking on "Deploy", it does not start the event handlers and beans are not initialized when they are registered in applicationLoad. For this to happen, I have to click on "Update" afterwards or trigger an application reload, which is not really clear in the user interface.
Some explanation would be useful, such as "You have updated the plugin files, but if this plugin registers an applicationLoad event, you also have to reload the application for the update to complete". It would be even better if the plugin could somehow be initialized without an application reload (especially when it is just for beans), but that might not be possible.

**Comment by David-Polehonski** [comment]
Yes, I am familiar with the latter issue having developed several plugins myself. This is due to a couple of things, firstly when you deploy a new zip folder it's extracted but the appropriate install/update functions aren't invoked until you accept the plugins license and any settings the plugin requires. Which makes sense but does require clear documenting in the plugins sections of the docs.

Secondly, your right the application isn't re-initialised when you install a plugin so beans and event listeners aren't registered, however this (I believe) is by design; not all plugins require an application reload. I prefer to address this issue by either manually adding the event handlers in the plugin.cfc install/update methods so they are available immediately as well as using on onApplicationLoad listener or I trigger the application reload after the install/update method has finished.
```
// plugin.cfc
component extends='mura.plugin.plugincfc' {
  public void function install() {
    // do some install stuff and then:
    variables.pluginConfig.addEventHandler(new path.to.plugin.eventHandler());

    // I suppose you could also do this if your registering non-transient beans:
    m.globalConfig().registerModelDir('/pluginName/path/to/your/model/');
  }

 public void function update() {
    /*
      Update stuff here and then reload application to ensure old handlers are removed
      and new ones are activated
    */
    applicationStop() // built in CF method for ending the application, triggering a reload
  }
}
```

**Comment by damien-git** [comment]
Adding that to the install handlers makes perfect sense, thanks for the tip !
I wonder if we could remove the old handlers without reloading the entire application, in the update method...

**Comment by David-Polehonski** [comment]
I can't see any way or removing eventhandlers, but there is a helpful design pattern you can use. Event Handler methods are resolved at the point they are registered, however any files they include are resolved at runtime, so if you write your event handler methods to just call an included file then you can change their behaviour at runtime (i.e on update) by replacing the content of the included file. 

eventHandler.cfc
```
component {
  public void function onSiteRequestStart(required struct mura) {
    include 'events/onSiteRequestStart.cfm';
  }
}
```
events/onSiteRequestStart.cfm (Before update)
```
  // Add custom scope at the start of the request to give access to plugin methods
  mura.setCustomMuraScopeKey('myPluginCustomScope', new plugin.customScope(mura));
```

events/onSiteRequestStart.cfm (After update)
```
  // Event logic has been removed from this file in version x.x.x handler won't be re-registered on reload.
```

**Comment by damien-git** [comment]
I did some experiments. When I click on "Deploy", the install and update methods are not called. When I click on "Update", the install or update method is called, but always followed by an application reload. So registering the beans in install is not useful as I already do it on application reload. There might be a way to disable application reload after an update (which might be triggered by application.appInitialized=false; in plugincfc.cfc), but since it is the default behavior I don't know if this is a good idea.

Using includes in events as you suggested should let me change them with a simple "Deploy". But could that have an impact on performance ?
